### ISSUE TITLE
Longer timeout limit for uploads

### ISSUE THREAD

**Issue by beroso** [781flyingdutchman/background_downloader#104]
On our application we noticed that the large files are failing to be uploaded. The failure exception looks like this:
`TaskException: Error for url https:/<my-server-route>/ and /data/user/0/<my-package-dir>/capture_1686911189424508.mp4: unexpected end of stream`

I suspect that we are facing the Android time limit constraint.
I see that on Android there is a time limit of 9 minutes, and for downloads we can set the `DowloadTask` `allowPause = true` to work around it.

But i not found an alternative for `UploadTask`.

**Comment by 781flyingdutchman** [comment]
Yes, it's likely you run into that 9 minute issue, and unfortunately there is no 'pause' option for uploads. What you need is a way for the background task to actually live more than 9 minutes, and that requires that background task to be run as a foreground service, which brings with it other requirements, so I have decided not to go down that route in the current implementation.

Is there any way you can chunk up the upload into smaller files? This would of course require you to control the server you're uploading to, but that would help and honestly would make the upload more robust in general: long running uploads are going to be sensitive to network interruptions etc. Or, if you control the server, can the server let you know how much it received up until the error, so you can choose to upload only part the missing part of the file in the next upload (instead of resending the entire file)?

**Comment by 781flyingdutchman** [comment]
Thanks for doing that. I've been thinking about this too, but am not quite clear on the implications/limitations of a foreground service so I'll experiment with your branch. I'm thinking if the concept works, I could choose a foreground worker when the file size exceeds X bytes, and let X be something a developer can set. That way, the default doesn't change, and the developer is in control about triggering it. I think I prefer that over adding another field to the Task, especially because this is an Android only feature/issue.
What do you think?

**Comment by beroso** [comment]
Awesome!

As this will be an Android only feature, I was wondering about a explicit config for it. Some suggestions:

Global config:
```dart
FileDownloader().platformConfig(
  android: AndroidSettings(
    supportForegroundWorkers: true,
  ),
);
```

Or an Task field:

```dart
UploadTask(
  ...
  androidSettings: AndroidSettings(
    supportForegroundWorkers: true,
  ),
);
```

**Comment by 781flyingdutchman** [comment]
I've added the notion of configuration, and the first configuration is for setting the Android 'runInForeground' mode, which should get around the time limitation and may solve the aggressive cancellation of backgroudn tasks mentioned in issue #89 as well.

I could really use your help testing this, both for uploads and downloads

Please update your pubspec.yaml to:
```
background_downloader:
    git:
      url: https://github.com/781flyingdutchman/background_downloader.git
      ref: config
```
Configuration is 'global', so once you configure something it stays in place until you change the config.

To configure the downloader to run everything in foreground mode, call:
```
    final configResult = await FileDownloader()
        .configure(androidConfig: ('runInForeground', true));
```
Note the argument to `.configure` is a record (and can be a list of records), hence the brackets.  The `configResult` is a list containing records where the first item is the configuration string used, followed by an empty string if all went well.

For every task running after this, you should now see several references to 'foreground' in the log, except the last notification update (for 'complete' or 'failed') should NOT be in foreground (because otherwise it disappears).

You can also configure it such that it runs in foreground only for uploads/downloads above a certain size (expressed in MegaBytes) as follows:

```
    final configResult = await FileDownloader()
        .configure(androidConfig: ('runInForegroundIfFileLargerThan', 10));
```

Whichever config you use last will apply.

Please test this as thoroughly as you can, and let me know what you find.

**Comment by beroso** [comment]
Thank you!
I will test it this weekend.

**Comment by 781flyingdutchman** [comment]
Forgot to mention an important point: you _must_ have a 'running' notification configured, as showing a notification while the foreground service is active is a pre-requisite set by WorkManager in Android.  If you don't configure a running notification then the task will execute normally, i.e. not in the foreground.

Also, the configuration you set is stored in SharedPreferences, so persists across Application start/stops.  Even if you remove the config and run the app again, the old config will still be there.

Happy testing!

**Comment by 781flyingdutchman** [comment]
I've added other configuration options and refactored how that works, so please make sure to pull the most recent from the `config` branch, and please read the `CONFIG.md` file (on that branch) to see what config options are available.

**Comment by 781flyingdutchman** [comment]
Hi @beroso wondering if you've had a chance to test the latest?

**Comment by 781flyingdutchman** [comment]
Included in V7.9.0 via 'runInForeground' configuration
